build.go

```diff
@@ -0,0 +1,355 @@
+
+		composerPhpIniPath, err := writeComposerPhpIni(logger, context)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		composerGlobalBin, err := runComposerGlobalIfRequired(logger, context, composerGlobalExec, path, composerPhpIniPath)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		if composerGlobalBin != "" {
+			path = strings.Join([]string{
+				composerGlobalBin,
+				path,
+			}, string(os.PathListSeparator))
+		}
+
+		workspaceVendorDir := filepath.Join(context.WorkingDir, "vendor")
+
+		if value, found := os.LookupEnv(ComposerVendorDir); found {
+			workspaceVendorDir = filepath.Join(context.WorkingDir, value)
+		}
+
+		composerPackagesLayer, layerVendorDir, err := runComposerInstall(
+			logger,
+			context,
+			composerInstallOptions,
+			composerPhpIniPath,
+			path,
+			composerInstallExec,
+			workspaceVendorDir,
+			calculator)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		logger.Process("Writing symlink to %s", workspaceVendorDir)
+
+		err = os.Symlink(layerVendorDir, workspaceVendorDir)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		if os.Getenv(BpLogLevel) == "DEBUG" {
+			logger.Debug.Action("listing files in %s:", layerVendorDir)
+			files, err := ioutil.ReadDir(layerVendorDir)
+			if err != nil {
+				return packit.BuildResult{}, err
+			}
+			for _, f := range files {
+				logger.Debug.Detail(fmt.Sprintf("- %s", f.Name()))
+			}
+		}
+
+		err = runCheckPlatformReqs(logger, checkPlatformReqsExec, context.WorkingDir, composerPhpIniPath, composerPackagesLayer.Path, path)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		return packit.BuildResult{
+			Layers: []packit.Layer{
+				composerPackagesLayer,
+			},
+		}, nil
+	}
+}
+
+func runComposerGlobalIfRequired(
+	logger scribe.Emitter,
+	context packit.BuildContext,
+	composerGlobalExec Executable,
+	path string,
+	composerPhpIniPath string) (string, error) {
+	composerInstallGlobal, found := os.LookupEnv(BpComposerInstallGlobal)
+
+	if !found {
+		return "", nil
+	}
+
+	logger.Process("Running 'composer global require'")
+
+	composerGlobalLayer, err := context.Layers.Get(ComposerGlobalLayerName)
+	if err != nil {
+		return "", err
+	}
+
+	composerGlobalLayer, err = composerGlobalLayer.Reset()
+	if err != nil {
+		return "", err
+	}
+
+	globalPackages := strings.Split(composerInstallGlobal, " ")
+
+	composerGlobalBuffer := bytes.NewBuffer(nil)
+
+	execution := pexec.Execution{
+		Args: append([]string{"global", "require", "--no-progress"}, globalPackages...),
+		Dir:  composerGlobalLayer.Path,
+		Env: append(os.Environ(),
+			"COMPOSER_NO_INTERACTION=1", // https://getcomposer.org/doc/03-cli.md#composer-no-interaction
+			fmt.Sprintf("COMPOSER_HOME=%s", composerGlobalLayer.Path),
+			fmt.Sprintf("PHPRC=%s", composerPhpIniPath),
+			"COMPOSER_VENDOR_DIR=vendor", // ensure default in the layer
+			fmt.Sprintf("PATH=%s", path),
```

Comment:
Why are all of these set in both the global case and in the normal case?

Reply:
Composer needs several of them to function. I think "normally" on a system such as a development laptop that it would rely on PHP to be installed first, but we don't make that assumption here.

`COMPOSER_NO_INTERACTION` is helpful to prevent any expectation of user interaction.